src/porepy/grids/grid.py

```diff
@@ -422,83 +423,94 @@ def nrm(u):
     def _compute_geometry_2d(self) -> None:
         """Auxiliary function to compute the geometry for 2D grids.
 
-        The method is motivated by MRST.
-
+        We assume that:
+        - either the cell_faces and face_nodes are consistently oriented
+        - or that the grid is composed of convex cells.
         """
 
-        R = pp.map_geometry.project_plane_matrix(self.nodes)
-        self.nodes = np.dot(R, self.nodes)
+        # Each face is determined by a start and end node, the tangent is given by
+        # the x_end - x_start. The face normal is a 90 degree clock-wise rotation
+        # of the tangent.
+
+        # Define an oriented face to nodes mapping, the orientation is determined by the
+        # ordering in self.face_nodes.indices. The start node gets a -1 and the end node a +1.
+        fn_orient = self.face_nodes.copy()
+        fn_orient.data = -np.power(-1, np.arange(fn_orient.data.size))
+
+        # Check consistency
+        is_oriented = (fn_orient * self.cell_faces).nnz == 0
+        if not is_oriented:
+            warnings.warn(
+                "Orientations in face_nodes and cell_faces are inconsistent, "
+                "we will now assume convex cells."
+            )
 
-        fn = self.face_nodes.indices
-        edge1 = fn[::2]
-        edge2 = fn[1::2]
-
-        xe1 = self.nodes[:, edge1]
-        xe2 = self.nodes[:, edge2]
-
-        edge_length_x = xe2[0] - xe1[0]
-        edge_length_y = xe2[1] - xe1[1]
-        edge_length_z = xe2[2] - xe1[2]
-        self.face_areas = np.sqrt(
-            np.power(edge_length_x, 2)
-            + np.power(edge_length_y, 2)
-            + np.power(edge_length_z, 2)
+        # Compute the tangent vectors and use them to compute face attributes
+        tangent = self.nodes * fn_orient
+        self.face_areas = np.sqrt(np.square(tangent).sum(axis=0))
+        self.face_centers = 0.5 * self.nodes * np.abs(fn_orient)
+
+        # Compute the temporary cell centers as average of the cell nodes
+        faceno, cellno, cf_orient = sps.find(self.cell_faces)
+        cx = np.bincount(cellno, weights=self.face_centers[0, faceno])
+        cy = np.bincount(cellno, weights=self.face_centers[1, faceno])
+        cz = np.bincount(cellno, weights=self.face_centers[2, faceno])
+        temp_cell_centers = np.vstack((cx, cy, cz)) / np.bincount(cellno)
+
+        # Create sub-simplexes based on (face, cell center) pairs.
```

Comment:
You mean a triplet of two face-nodes and one cell center?

Reply:
Yes, it is the simplex formed by the two face-nodes and the cell center